### ISSUE TITLE
Add unique page view to each blog using Google analytics API

### ISSUE THREAD

**Issue by poanchen** [poanchen/poanchen.github.io#13]
Add page views to each blog post, data from Google analytics (https://developers.google.com/analytics/devguides/reporting/core/v4/quickstart/web-js)

**Comment by poanchen** [comment]
Looking into it a bit more in depth. Found that in order to access Google public data like map, youtube, you can do it through api key. In order to access private data like Google analytics, you need to use the Google oauth2 token which should only be accessible by me as I am the owner to the Google analytics according to this article https://stackoverflow.com/questions/41176535/how-to-get-google-analytics-data-using-simple-api-key-in-c

so far I have done this,
```
<!-- to be put somewhere else probably -->
<script src="https://apis.google.com/js/platform.js?onload=init" async defer></script>
<script>
  function init() {
    gapi.load('auth2', function() {
      gapi.auth2.init({
        client_id: 'client_id.apps.googleusercontent.com'
      }).then(function() {
        gapi.auth2.getAuthInstance().signIn().then(function() {
          var user = gapi.auth2.getAuthInstance().currentUser.get();
          console.log(user.getAuthResponse());
        });
      }, function(err) {
        console.log(err);
      })
    });
  }
</script>
```
In the end of this code, I am able to get my own access token for the Google analytics but it expiry in certain time.

[gapi.auth2.init doc](https://developers.google.com/identity/sign-in/web/reference#googleauththenoninit-onerror)
[google signin](https://developers.google.com/identity/sign-in/web/sign-in#before_you_begin)
[discovery google analytics api](https://content.googleapis.com/discovery/v1/apis/analyticsreporting/v4/rest)

batchget example
```
POST https://analyticsreporting.googleapis.com/v4/reports:batchGet?key=
Body
{
 "reportRequests": [
  {
   "viewId": "ga:XXX",
   "dimensions": [
    {
     "name": "ga:pagePath"
    }
   ],
   "metrics": [
    {
     "expression": "ga:pageviews"
    }
   ],
   "dimensionFilterClauses": [
    {
     "filters": [
      {
       "operator": "EXACT",
       "dimensionName": "ga:pagePath",
       "expressions": [
        "/blog/2016/11/20/how-to-enable-cross-origin-resource-sharing-on-an-apache-server"
       ]
      }
     ]
    }
   ],
   "dateRanges": [
    {
     "startDate": "2016-10-16",
     "endDate": "2018-03-13"
    }
   ]
  }
 ]
}
```
[batchget ref](https://developers.google.com/analytics/devguides/reporting/core/v4/rest/v4/reports/batchGet)
pretty cool if you need to explore how api works before working on it
[google analytics query explore](https://ga-dev-tools.appspot.com/query-explorer/)
[create google credital here](https://console.developers.google.com/apis/credentials)

**Comment by poanchen** [comment]
Forgot that I created this issue haha. Thanks for the investigation. Came back did the investigation again. In #21, we went with different approach. Instead of having to call the api in client side, we called it in server side. (this way all our secret is stored in server side should be safe as long as access to server is safe) Have a cron job around it. Upload the result to Azure Blob Storage. In client side, simply do GET to access the JSON file in the Azure Blob Storage. Tada. This feature is live on prod now.
